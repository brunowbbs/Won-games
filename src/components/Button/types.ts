import { AnchorHTMLAttributes, ButtonHTMLAttributes } from "react";

export type ButtonProps = {
  as?: React.ElementType;
  //children?: React.ReactNode;
  size?: "small" | "medium" | "large";
  fullWidth?: boolean;
  icon?: React.ReactNode;
  //onClick?: () => (event: React.MouseEvent<HTMLButtonElement>) => void;
} & ButtonTypes;

export type ButtonTypes =
  | AnchorHTMLAttributes<HTMLAnchorElement>
  | ButtonHTMLAttributes<HTMLButtonElement>;
