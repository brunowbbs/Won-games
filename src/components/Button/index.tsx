import * as S from "./styles";
import { ButtonProps } from "./types";

const Button = ({
  children,
  icon,
  size = "medium",
  fullWidth = false,
}: ButtonProps) => (
  <S.Wrapper size={size} fullWidth={fullWidth} hasIcon={!!icon}>
    {!!icon && icon}
    {!!children && <span>{children}</span>}
  </S.Wrapper>
);

export default Button;
