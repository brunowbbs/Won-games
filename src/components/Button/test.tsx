import { screen } from "@testing-library/react";
import { renderWithTheme } from "../../utils/tests/helpers";
import { FaBeer } from "react-icons/fa";
import "jest-styled-components";

import Button from ".";

describe("<Button/>", () => {
  it("Should render the medium size by default", () => {
    const { container } = renderWithTheme(<Button>Buy Now</Button>);
    expect(screen.getByRole("button", { name: /Buy now/i })).toHaveStyle({
      height: "4rem",
      padding: "0.8rem 3.2rem",
      "font-size": "1.4rem",
    });
    expect(container.firstChild).toMatchSnapshot();
  });

  it("Should render the small size", () => {
    renderWithTheme(<Button size="small">Buy Now</Button>);
    expect(screen.getByRole("button", { name: /Buy now/i })).toHaveStyle({
      height: "3rem",
      "font-size": "1.2rem",
    });
  });

  it("Should render the large size", () => {
    renderWithTheme(
      <Button icon={<FaBeer data-testid="icon" />}>Buy Now</Button>
    );
    expect(
      screen.getByRole("button", { name: /Buy now/i })
    ).toBeInTheDocument();
    expect(screen.getByTestId("icon")).toBeInTheDocument();
  });

  it("Should render a fullwidth version", () => {
    renderWithTheme(<Button fullWidth>Buy Now</Button>);
    expect(screen.getByRole("button", { name: /Buy now/i })).toHaveStyle({
      width: "100%",
    });
  });

  it("Should render an icon version", () => {
    renderWithTheme(<Button fullWidth>Buy Now</Button>);
    expect(screen.getByRole("button", { name: /Buy now/i })).toHaveStyle({
      width: "100%",
    });
  });

  it("Should render button has a link", () => {
    renderWithTheme(
      <Button as="a" href="/link">
        Buy Now
      </Button>
    );

    expect(screen.getByRole("link", { name: /Buy now/i })).toHaveAttribute(
      "href",
      "/link"
    );
  });

  it("Should render a minimal version", () => {
    renderWithTheme(
      <Button size="small" minimal>
        Buy Now
      </Button>
    );

    expect(screen.getByRole("button", { name: /buy now/i })).toHaveStyle({
      backgroundColor: "none",
      color: "#F231A5",
    });

    expect(screen.getByRole("button", { name: /buy now/i })).toHaveStyleRule(
      "background",
      "none",
      { modifier: ":hover" }
    );
  });
});
