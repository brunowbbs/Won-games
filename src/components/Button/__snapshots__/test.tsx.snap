// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button/> Should render the medium size by default 1`] = `
<button
  class="sc-bcXHqe sXlZP"
>
  <span>
    Buy Now
  </span>
</button>
`;
